src/ir/proto/sql/decode.h

```diff
@@ -25,6 +25,9 @@
 
 namespace raksha::ir::proto::sql {
 
+const Value &DecodeExpression(
+    const Expression &expr, DecoderContext &decoder_context);
+
```

Comment:
This comment is not related to this PR, but why can't we return an `Operation` for `DecodeSourceTableColumnXYZ` methods? Then we can wrap them in a `Value` as needed?

Reply:
I'm using a `Storage` to represent the column, as that seemed like a pretty natural way to represent the source of the column (clearly a more mature system would represent the table with the `Storage` and use `Operation`s to grab columns out of that table). Once I chose a `Storage`, a `StoredValue` seemed like the natural way to represent the value coming out of that `Storage`.

Your question makes me wonder if `StoredValue` is necessary. If we introduced a new `StorageAttribute`, we could introduce a `load` `Operation`, hang the `Storage` from which it `load`ed onto the attribute, and just use the result of that `OperationResult` instead of `StoredValue`.

The way that we are leaning more and more on `Attribute`s is making me wonder if `Operator`s should further have on them a list of `expected_attribute`s, indicating the `Attribute`s that should be present upon an `Operation` associated with that `Operator` and the types we expect them to have. This could be useful for writing datalog rules to handle all `Operation`s associated with a particular `Operator` without constantly testing for `Attribute` presence.

Reply:
Yes, the `Operator` should have metadata like `expected_attributes` and `expected_inputs`. It should not change the current `Operation` API and will mostly serve as a way to validate the instances of operations. In my initial version of the `Operator`, I did have some metadata but removed them as there were some unanswered questions and wanted to think it through. e.g. how do we model variadic arguments like `Merge` operation and so on.

Yes, it looks like we can get rid of `StoredValue` as well. We could even get rid of ANY and make it an operation. Then value will be quite simple  as it is either a block arg or an operation result.

Reply:
Just to be clear, let us do this in  follow up PRs.

Reply:
Sounds good, resolving.